pkg/sentry/fsimpl/devpts/line_discipline.go

```diff
@@ -455,6 +456,80 @@ func (*inputQueueTransformer) transform(l *lineDiscipline, q *queue, buf []byte)
 			l.terminal.replicaKTTY.SignalForegroundProcessGroup(kernel.SignalInfoPriv(linux.SIGQUIT))
 		}
 
+		// In canonical mode, some characters need to be handled specially; for example, backspace.
+		if l.termios.LEnabled(linux.ICANON) {
+			// n_tty.c:n_tty_receive_char_canon
+			c := cBytes[0]
+			c_cc := l.termios.ControlCharacters
+
+			// c == c_cc[linux.VKILL] is also handled by n_tty.c:eraser, but this isn't implemented
+			if c == c_cc[linux.VERASE] || (c == c_cc[linux.VWERASE] && l.termios.LEnabled(linux.IEXTEN)) {
+				// n_tty.c:eraser
+				var kill_type int
+				if c == c_cc[linux.VERASE] {
+					kill_type = linux.VERASE
+				} else {
+					kill_type = linux.VWERASE
+				}
+				seen_alnums := 0
+				for len(q.readBuf) > 0 {
+					// Erase a single possibly multibyte character
+					var to_erase byte
+					cnt := 0
+					is_continuation := true
+					for {
+						to_erase = q.readBuf[len(q.readBuf)-cnt-1]
+						cnt++
+						is_continuation = l.termios.IEnabled(linux.IUTF8) && (to_erase&0xc0) == 0x80
+						if cnt == len(q.readBuf) || !is_continuation {
+							break
+						}
+					}
+					if is_continuation {
+						// do not partially erase
+						break
+					}
+
+					if kill_type == linux.VWERASE {
+						if unicode.IsLetter(rune(to_erase)) || unicode.IsDigit(rune(to_erase)) || to_erase == '_' {
+							seen_alnums++
+						} else if seen_alnums > 0 {
+							break
```

Comment:
Why break here when seenAlnums is > 0?

Reply:
I think this is what the Linux kernel did. My understanding is VWERASE will stop erasing characters if the character it's about to erase is not alphanumeric, and we've deleted at least one alphanumeric character already.